alembic/script/revision.py

```diff
@@ -156,12 +185,51 @@ def _revision_map(self):
                 down_revision = map_[downrev]
                 down_revision.add_nextrev(rev)
                 if downrev in rev._versioned_down_revisions:
-                    heads.discard(downrev)
-                _real_heads.discard(downrev)
+                    heads.discard(down_revision)
+                _real_heads.discard(down_revision)
+
+        if rev_map:
+            if not heads or not bases:
+                raise CycleDetected
+            total_space = set(
```

Comment:
**mike bayer** (zzzeek) wrote:

can we use set/iterator comprehensions for these, I find them easier to read:

    total_space = {
        rev.revision
        for rev in self._iterate_related_revisions(
            lambda r: r._versioned_down_revisions, heads, map_=rev_map
        )
    }.intersection(
        rev.revision
        for rev in self._iterate_related_revisions(
            lambda r: r.nextrev, bases, map_=rev_map
        )
    )

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2379